compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -168,8 +168,48 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
         mplace: &MPlaceTy<'tcx>,
         fields: impl Iterator<Item = InterpResult<'tcx, Self::V>>,
     ) -> InterpResult<'tcx> {
-        // ZSTs cannot contain pointers, so we can skip them.
-        if mplace.layout.is_zst() {
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
+        // also need to take special care of interior mutability.
+        //
+        // As an optimization, however, if the allocation does not contain any pointers: we don't
+        // need to do the walk. It can be costly for big arrays for example (e.g. issue #93215).
+        let is_walk_needed = |mplace: &MPlaceTy<'tcx>| -> InterpResult<'tcx, bool> {
+            // ZSTs cannot contain pointers, we can avoid the interning walk.
+            if mplace.layout.is_zst() {
+                return Ok(false);
+            }
+
+            // Now, check whether this allocation contains reference types (as relocations).
```

Comment:
```suggestion
            // Now, check whether this allocation contains pointers (as relocations).
```
This is not about the types, it's about the actual values.

Reply:
I will change this but I was copying from the few lines below which contain both "references" and "reference types" https://github.com/rust-lang/rust/blob/94e93749ab00539a11e90426ea87382c433530a8/compiler/rustc_const_eval/src/interpret/intern.rs#L176-L194 should I change something else in the file ?

Reply:
It does not contain "reference types".

But yes it does say "references" and... indeed the internet only looks for references, not raw pointers. So let's agree on "references"?

Reply:
It does at the bottom, here it as as a single link so that it's not hidden by the scrollbar: https://github.com/rust-lang/rust/blob/94e93749ab00539a11e90426ea87382c433530a8/compiler/rustc_const_eval/src/interpret/intern.rs#L192-L195

Reply:
Ah yes, but that code is talking about `ty::Ref` -- reference types.

Reply:
I can change all the "pointers" to "references" if you'd prefer (or the opposite) ? I still need to launch another perf run after this one anyways, to see if the original commit shows change on `diesel` and could do that at the same time.

Reply:
The thing is, "whether this allocation contains reference types (as relocations)" is meaningless. Relocations represent reference (and pointer) *values*. And it is the (reference) *values* that we are after here, the types just guide us there. But there being no relocations does *not* mean there there are no reference types! It just means that there are no reference *values*.

Reply:
Ah, I only just saw you already changed that  -- sorry.

So, yeah, it should probably say references in some of the places it currently says pointers. But at least some of the existing code is right when it talks about pointers... so if you change it, only change it in comments you added.

Reply:
Will do. 

I'll also add clarifications to the other comments I added elsewhere, to match your latest comment which was quite clear. Thanks for taking the time to explain.

Reply:
To ensure I understood correctly (it's late here), @RalfJung, IIUC the following would be acceptable, right ?

> Now, check whether this allocation could contain references.
>
> If there are no relocations in this allocation, it does not contain references and we can avoid the interning walk.

Reply:
Seems fine... if I am super nitpicky we'd have to go into detail about pointer and integer values and how values of reference types can be either, but maybe let's not. ;)

Maybe "does not contain references that point to another allocation"? That's hopefully still understandable and also technically correct. ;)

(The concern is references like `&*(1usize as *const ())`, which is a perfectly fine value of type `&()` but does not involve a relocation.)

Reply:
I've pushed a commit to address this.